### PR TITLE
config: add exec-shutdown for running commands on shutdown

### PR DESCRIPTION
**Pull request by Trimutex**
#### Describe your PR, what does it fix/add?
Adds a config option for executing programs on shutdown.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
Nothing that could crash Hyprland while running immediately found.

#### Is it ready for merging, or does it need work?
Ready, but want tested by #7671 for any edge cases I might be able to fix

### CODE REVIEW

**Comment by littleblack111** [comment]
@Trimutex I tried it. doesnt really work(I tried hyprctl clients > ~/asdf, which reads "Couldn't read (5)". I think `wl_display_terminate` prob already terminated all the windows? I tried to move `g_pConfigManager->dispatchExecShutdown();` to `CCompositor::stopCompositor` right before `wl_display_terminate` but it doesnt seem to execute then. I tried to just putting logs(`Debug::log(INFO, "stuff"`) but the whole `stopCompositor` part seem to not execute any of the code inside. uncertain how does this work(cuz im noob to coding wayland) but it seems to be the first function that get called when its terminating(by signal).

But anyways. I don't know how would I continue this... But, I hope we can add a exec-pre-shutdown and a exec-post-shutdown. the pre one for first immediate action before any window gets destroyed or anything(e.g. to save info like hyprctl clients, but im not sure if this not working is becuz the IPC shuts down?). But thank you for giving it a go! Hope we can find a solution soon

File: src/Compositor.cpp [hunk 1]
```diff
@@ -460,6 +460,8 @@ void CCompositor::cleanup() {
     m_bIsShuttingDown   = true;
     Debug::shuttingDown = true;
 
+    g_pConfigManager->dispatchExecShutdown();
```
Comment (vaxerski) [inline src/Compositor.cpp:463]: should be executed just before signal() a few lines up

Comment (Trimutex) [reply src/Compositor.cpp]: Execs were started too late anyways, resolved.

**Comment by Trimutex** [comment]
@littleblack111 this should provide what you are looking for now I think

**Review by vaxerski — APPROVED** [review]

**Comment by littleblack111** [comment]
@Trimutex is there any way to make it execute it when is shutdown not via hyprctl dispatch exit? I tried sigterm, it doenst execute it.

**Comment by Trimutex** [comment]
I'm a little confused about this. Are you asking for it to be handled post-hyprland-shutdown or as its resetting the variables?

**Comment by littleblack111** [comment]
so, if I shut off hyprland via the the exit dispatcher. it will execute. but any other way(reboot, soft-reboot, restart display manager) will not trigger it.

**Comment by Trimutex** [comment]
> so, if I shut off hyprland via the the exit dispatcher. it will execute. but any other way(reboot, soft-reboot, restart display manager) will not trigger it.

The closest you can come to that is a wrapper for Hyprland in a script I think. This seems like a bit of an XY problem to me tho

**Comment by Trimutex** [comment]
If you want stuff to run on reboot or similar a bash script attached to a restart button inside of Hyprland that is run which calls reboot at end of script is my best guess of what you actually want

**Comment by littleblack111** [comment]
as described at #7671 I want to do #2475 and possibly pause the shutdown or something if some important task is being executed

**Comment by Trimutex** [comment]
> as described at #7671 I want to do #2475 and possibly pause the shutdown or something if some important task is being executed

I don't think Hyprland itself is capable of stopping reboot for updates. You can defer the reboot/shutdown call though by using a script that is called to replace shutdown by whatever method you prefer.

**Comment by littleblack111** [comment]
is the shutdown process if linux un-pausable?

**Comment by Trimutex** [comment]
> is the shutdown process if linux un-pausable?

theoretically yes but realistically only if its scheduled for later. attempting to pause it for updates seems unnecessary anyways as other solutions make more sense (e.g. wrapper script for shutdown that checks if a file contains a "1", skipping shutdown if it does)

**Pull request merged**